spec/inputs/exec_spec.rb

```diff
@@ -3,75 +3,118 @@
+
+    before(:each) do
+      allow_any_instance_of(described_class).to receive(:ecs_compatibility).and_return(ecs_compatibility)
     end
-    context "with an invalid config" do
-      let(:input) { described_class.new("command" => "ls") }
-      it "should raise error" do
-        expect {input.register}.to raise_error(LogStash::ConfigurationError)
+
+    context "when register" do
+      let(:input) { described_class.new("command" => "ls", "interval" => 0) }
+
+      it "should not raise error if config is valid" do
+        # register will try to load jars and raise if it cannot find jars or if org.apache.log4j.spi.LoggingEvent class is not present
+        expect { input.register }.to_not raise_error
       end
-    end
-  end
 
-  context "when operating normally" do
-    let(:input) { described_class.new("command" => "ls", "interval" => 0) }
-    let(:queue) { [] }
-    let(:loggr) { double('loggr') }
-
-    before :each do
-      expect(described_class).to receive(:logger).and_return(loggr).exactly(7).times
-      allow(loggr).to receive(:info)
-      allow(loggr).to receive(:info?)
-      allow(loggr).to receive(:warn)
-      allow(loggr).to receive(:warn?)
-      allow(loggr).to receive(:debug)
-      allow(loggr).to receive(:debug?)
+      context "with an invalid config" do
+        let(:input) { described_class.new("command" => "ls") }
+        it "should raise error" do
+          expect { input.register }.to raise_error(LogStash::ConfigurationError)
+        end
+      end
     end
 
-    it "enqueues some events" do
-      input.register
-      expect(loggr).not_to receive(:error)
+    context "when operating normally" do
+      let(:input) { described_class.new("command" => "echo 'hi!'", "interval" => 0) }
+      let(:queue) { [] }
 
-      input.execute(queue)
+      before :each do
+        input.register
+      end
 
-      expect(queue.size).not_to be_zero
-    end
-  end
+      it "enqueues some events" do
+        expect(input.logger).not_to receive(:error)
 
-  context "when a command runs normally" do
-    let(:input) { described_class.new("command" => "/bin/sh -c 'sleep 1; /bin/echo -n two; exit 3'", "interval" => 0) }
-    let(:queue) { [] }
+        input.execute(queue)
 
-    before do
-      input.register
-      input.execute(queue)
+        expect(queue.size).not_to be_zero
+      end
     end
 
-    after do
-      input.stop
-    end
+    context "when command fails" do
+      let(:input) { described_class.new("command" => "invalid_command 1 2 3", "interval" => 0) }
+      let(:queue) { [] }
+
+      before :each do
+        input.register
+      end
+
+      it "does not enqueue an event (in a non-Docker env)" do
+        expect(input.logger).to receive(:error).and_call_original
+
+        input.execute(queue)
+
+        expect(queue.map(&:to_hash)).to be_empty
+      end
+    end if ENV['CI'] != 'true' # in Docker the behavior differs - missing command files are not raised
```

Comment:
oof. I don't like skipping specs in CI. Can you explain why this is needed? what doesn't raise?

Reply:
sorry I forgot about this one - should have mentioned the behavior difference in Docker vs non-Docker.
normally one gets `Errno::ENOENT: No such file or directory - invalid_command` for a missing command,
however the behavior is different in Docker - `popen` does not raise - we simply get a non-zero exit code:
`"message"=>"", "process"=>{"command_line"=>"invalid_command 1 2 3", "exit_code"=>127}`

haven't looked into this one as it's existing behavior - thought we rather have this as a known behavior spec-ed.

Reply:
If there are two different behaviours, depending on which platform it is run, I would prefer we clearly specify both behaviours instead of specifying one of the behaviours and conditionally executing that spec.

If possible, it could be helpful to _emulate_ the underlying change in behaviour and define our specs so that we can validate our effective behavour in CI regardless of which platform it is executed on (It would be hard to stub `IO::popen` because we rely on its side-effect of setting `$?`, but can we stub `run_command`?).

So our specs output would look something like:

 - `when command fails`
   - `on a platform where IO.popen raises ENOENT`
     - `it does not enqueue an event`
   - `on a platform where IO.popen does not raise ENOENT (Docker)`
     - `it enqueues an event tagged with the error code`

Reply:
:+1: if we're going this far as to specify the behavior that I would rather move this out of this PR.
Have opened an issue https://github.com/logstash-plugins/logstash-input-exec/issues/29 + I will create a draft with the suggested spec. The problematic (newly added) spec has been removed.